Constexpr Tensor

rezarastak opened the pull request:
Following #8011 and being careful about #8037, this is the second step towards #8003.
1. Defined the macro `DEAL_II_CONSTEXPR` that represents `constexpr` in c++14-enabled compilers and disappears in older compilers with c++11. Let me know if I defined it the correct way.
1. Added `constexpr` to simple functions and `DEAL_II_CONSTEXPR` to complicated functions in `tensor.h` and `tensor_accessors.h`.
1. Removed `inline` from functions that are `constexpr` because `constexpr` implies `inline` based on the standard.
1.  Removed `DEAL_II_ALWAYS_INLINE` from functions that are `constexpr`. My reason is that any function that has a single return statement (satisfies c++11 `constexpr` requirements) will be inlined by the compiler in almost all circumstances.
1. Functions decorated with `DEAL_II_CONSTEXPR` still include `inline` and `DEAL_II_ALWAYS_INLINE` attributes.
1. Changed the constructors for `Tensor<0, dim, Number>` and `Tensor<rank, dim, Number>`. The constructors are now `constexpr` (c++11). I had to use variadic templates to make it work. It looks relatively elegant but I needed to add a private helper constructor to `Tensor<0, dim, Number>`.
1. Because of the aforementioned changes to the constructors, I had to change the constructors of `VectorizedArray<Number>` as well.

The result of all of these changes is that we can compile the following code in a portable way and make use `constexpr` features.
```
constexpr double tensor_0_dim_constexpr(){
  constexpr Tensor<0,2> a{7.6};
  return a;
}

DEAL_II_CONSTEXPR double tensor_2_dim_calc(){
  DEAL_II_CONSTEXPR const double a[3][3] =
    {{4., 2., 0.}, {2., 3., 0.},{0., 0., 5.}};
  DEAL_II_CONSTEXPR const Tensor<2,3> A(a);
  DEAL_II_CONSTEXPR const auto det =  determinant(A);  // 40
  DEAL_II_CONSTEXPR const auto tr = trace(transpose(A) * A); // 58
  return det + tr;
}

void use_constexpr_functions(){
  constexpr auto x = tensor_0_dim_constexpr();   // returns 7.6
  DEAL_II_CONSTEXPR auto y = tensor_2_dim_calc();  // returns 98.0
}
```
If c++14 is enabled, the assembly code created for the function `use_constexpr_functions()` in debug mode is:
```
   1dcaa:       55                      push   %rbp
   1dcab:       48 89 e5                mov    %rsp,%rbp
   1dcae:       c7 45 f8 07 00 00 00    movl   $0x7,-0x8(%rbp)
   1dcb5:       c7 45 fc 62 00 00 00    movl   $0x62,-0x4(%rbp)
   1dcbc:       90                      nop
   1dcbd:       5d                      pop    %rbp
   1dcbe:       c3                      retq   
```
All the complicated tensor multiply and determinant operations where performed at compile time.

kronbichler commented on include/deal.II/base/vectorization.h:
```diff
@@ -175,8 +175,23 @@ class VectorizedArray
+   */
+  explicit constexpr VectorizedArray(const Number scalar)
+    : data{scalar}
+  {}
```
Have you checked what `std::is_pod<VectorizedArray<double>>::value` returns after this change, both for this default class and the SIMD classes below? If this does not succeed, we could probably relax the 'no constructors' design of this class anyway, because it would simplify code in a number of places. It would be good if we could at least guarantee `std::is_standard_layout` as that assumption is used implicitly all over the place (not that I would have seen any sane compiler doing it wrong also without that yet). A follow-up question would be to what extent we could make this class `constexpr` in a similar vein as `Tensor`, but that can wait for another PR for sure.

rezarastak replied:
I removed unnecessary constructors from VectorizedArray<Number>, but these constructors are still defined for the specializations of the class. I added static_assert for make sure the class has standard layout.

Rombur commented:
Before we merge this, I want to check that it works with CUDA. I don't know how good `constexpr` support is with `nvcc`

drwells commented:
Quick comment: the point of `ALWAYS_INLINE` is not to specify `inline` but to use a compiler-specific pragma to force actual function inlining. We should keep this regardless of the `constexpr` status of the function.

drwells commented:
It may be overly paranoid to keep but we did add those annotations for a reason :)

masterleinad commented:
> Before we merge this, I want to check that it works with CUDA. I don't know how good `constexpr` support is with `nvcc`

For me, all the `CUDA` tests are passing with `CUDA-8`.

masterleinad commented:
Given our experience with #8034, we should also test with old compilers like `gcc-4.8`.

Rombur commented:
> For me, all the CUDA tests are passing with CUDA-8.

CUDA 8 does not support C++14 so the PR should not use `constexpr`

masterleinad commented on cmake/checks/check_01_cxx_features.cmake:
```diff
@@ -501,6 +502,16 @@ OPTION(DEAL_II_WITH_CXX17
+  MESSAGE(STATUS "Enabling CXX14 constexpr features")
+  SET(DEAL_II_CONSTEXPR "constexpr")
+ELSE()
+  SET(DEAL_II_CONSTEXPR " ")
```
If you replace `inline` with `constexpr` you are effectively removing `inline` when compiling without `C++14` support. That is likely not intended.

rezarastak replied:
The functions I decorate with `constexpr` are one-liners that will be constexpr even in old c++11 compilers.

All other functions are decorated with `DEAL_II_CONSTEXPR inline` which will remain inline both with or without c++14 support. Some of these functions are additionally decorated with `DEAL_II_ALWAYS_INLINE`.

So there are no portability issues with regard to the definition of `DEAL_II_CONSTEXPR`.

bangerth commented:
Could I suggest that we postpone merging this until after the release? We're unlikely to get anything complete in terms of `constexpr` into the library by the deadline today, and the risk is substantial that we break some compiler or other.

I like the direction of these patches, but let's wait until after the branch has been created.

rezarastak commented:
> Given our experience with #8034, we should also test with old compilers like `gcc-4.8`.

@masterleinad I have compiled it together with MPI, TRILINOS, ADOL-C, SACADO using
```
$ mpicxx --version
g++ (Ubuntu 4.8.4-2ubuntu1~14.04.4) 4.8.4
```
and
```
$ mpicxx --version
g++ (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
```
without any errors or warnings.

bangerth commented:
On 5/10/19 11:29 AM, Reza Rastak wrote:
> So there are no portability issues with regard to the definition of 
> |DEAL_II_CONSTEXPR|.

No no, we know this isn't true :-) Just because something *ought* to 
work doesn't mean that it *will* work given the diversity of all of the 
compilers out there :-)

rezarastak commented:
@bangerth I totally agree :-)
I can (mostly!) guarantee that my implementation is portable across c++ standards (11, 14, 17, 20) and the two compilers that I have, but we have to wait and ensure that it works for all supported compilers (after the 9.1 release of course).

masterleinad commented:
With newer `CUDA` versions we at least get more warnings like
```
/export/home/darndt/dealii-cuda-10/include/deal.II/base/tensor.h(779): warning: calling a constexpr __host__ function("Tensor") from a __host__ __device__ function("Tensor") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
```

masterleinad commented:
> >On 5/10/19 11:29 AM, Reza Rastak wrote: So there are no portability issues with regard to the definition of |DEAL_II_CONSTEXPR|.
>
> No no, we know this isn't true :-) Just because something *ought* to work doesn't mean that it *will* work given the diversity of all of the compilers out there :-)

Adding to this: Just because a compiler accepts a certain version flag, does not mean that they are fully compliant and implement all features. That is why we typically test all the `C++14` and `C++17` features we use at configure time. We might want to do something similar here as well.

rezarastak commented:
Now that the 9.1 release has its own branch and will not be affected by this PR, I did some additional work on this. I added a test for the `constexpr` constructors of `Tensor`. I also added a minor changelog.


> Quick comment: the point of `ALWAYS_INLINE` is not to specify `inline` but to use a compiler-specific pragma to force actual function inlining. We should keep this regardless of the `constexpr` status of the function.

@drwells  I understand your concern, so I added `DEAL_II_ALWAYS_INLINE` back to the previous places. I am curious to know how we can measure the benefits of this macro. I also wonder if there has been any compiler that did not inline a one-line constexpr function.

> Adding to this: Just because a compiler accepts a certain version flag, does not mean that they are fully compliant and implement all features. That is why we typically test all the `C++14` and `C++17` features we use at configure time. We might want to do something similar here as well.

@masterleinad  I changed the configuration. Now `DEAL_II_CONSTEXPR` is activated only when the macro `DEAL_II_HAVE_CXX14_CONSTEXPR_CAN_CALL_NONCONSTEXPR` is true. This makes sure that the compiler supports all `consexpr` features that I am using. It is slightly conservative.



> With newer `CUDA` versions we at least get more warnings like
> 
> ```
> /export/home/darndt/dealii-cuda-10/include/deal.II/base/tensor.h(779): warning: calling a constexpr __host__ function("Tensor") from a __host__ __device__ function("Tensor") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
> ```

What do you think is the best approach to deal with this warning? Is there any disadvantage with adding the flag `--expt-relaxed-constexpr`?

masterleinad commented:
> What do you think is the best approach to deal with this warning? Is there any disadvantage with adding the flag `--expt-relaxed-constexpr`?

No, we should be fine since we require `CUDA 8` or later.

masterleinad commented on include/deal.II/base/vectorization.h:
```diff
@@ -782,7 +782,7 @@ class VectorizedArray<double>
    * remains public.
    */
-  __m512d data;
+  __m512d data{0.};
```
Is there any reason to no just do 
```suggestion
  __m512d data = _mm512_set1_pd(0.);
```
instead of defining a conversion constructor and an explicit default constructor, and similarly below?

masterleinad commented on include/deal.II/base/vectorization.h:
```diff
@@ -379,7 +367,7 @@ class VectorizedArray
    * declared public.
    */
-  Number data;
+  Number data{0.};
```
```suggestion
  Number data = 0.;
```

masterleinad commented:
/rebuild

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -18,8 +18,13 @@
+constexpr unsigned int VectorizedArray<float>::n_array_elements;
+static_assert(
+  std::is_standard_layout<VectorizedArray<double>>::value,
+  "VectorizedArray<double> must has standard (C compatible) layout");
```
```suggestion
  "VectorizedArray<double> must have standard (C compatible) layout");
```

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -18,8 +18,13 @@
+  std::is_standard_layout<VectorizedArray<double>>::value,
+  "VectorizedArray<double> must has standard (C compatible) layout");
+static_assert(std::is_standard_layout<VectorizedArray<float>>::value,
+              "VectorizedArray<float> must has standard (C compatible) layout");
```
```suggestion
              "VectorizedArray<float> must have standard (C compatible) layout");
```

masterleinad commented on tests/base/constexpr_tensor.cc:
```diff
@@ -0,0 +1,70 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2005 - 2018 by the deal.II authors
```
```suggestion
// Copyright (C) 2019 by the deal.II authors
```

masterleinad commented:
There are a number of test failures:
- `tests/base/vectorization_01.cc:34`, `tests/base/vectorization_06.cc:43`
```
error: no match for ‘operator=’ (operand types are ‘dealii::VectorizedArray<long double>’ and ‘long double’)
```
- `tests/base/vectorization_09.cc:38`, `tests/base/vectorization_04.cc:32`
```
error: narrowing conversion of ‘0.0’ from ‘double’ to ‘unsigned int’ inside { } [-Wnarrowing]
```
- `base/vectorization_07` checks that the class is `pod` and `trivial`. This test can likely just be removed.

There are also some complains about unused variables in the `SymmetricTensor` headers. I'll fix that in a separate pull request.

masterleinad commented:
> constexpr double tensor_0_dim_constexpr(){
>   constexpr Tensor<0,2> a{7.6};
>   return a;
> }
> 
> DEAL_II_CONSTEXPR double tensor_2_dim_calc(){
>   DEAL_II_CONSTEXPR const double a[3][3] =
>     {{4., 2., 0.}, {2., 3., 0.},{0., 0., 5.}};
>   DEAL_II_CONSTEXPR const Tensor<2,3> A(a);
>   DEAL_II_CONSTEXPR const auto det =  determinant(A);  // 40
>   DEAL_II_CONSTEXPR const auto tr = trace(transpose(A) * A); // 58
>   return det + tr;
> }
> 
> void use_constexpr_functions(){
>   constexpr auto x = tensor_0_dim_constexpr();   // returns 7.6
>   DEAL_II_CONSTEXPR auto y = tensor_2_dim_calc();  // returns 98.0
> }

That would make for a nice test.

rezarastak commented on include/deal.II/base/vectorization.h:
```diff
@@ -178,21 +178,9 @@ class VectorizedArray
    * can work on multiple elements at the same time.
    */
-  static const unsigned int n_array_elements = 1;
+  static constexpr unsigned int n_array_elements = 1;
```
Yes it doesn't change anything for integral types. I can revert it.

rezarastak replied:
Changed it back to const.

rezarastak commented on include/deal.II/base/vectorization.h:
```diff
@@ -588,7 +576,19 @@ class VectorizedArray<double>
+   * Conversion constructor from a double
+   */
+  DEAL_II_ALWAYS_INLINE explicit VectorizedArray(const double x)
+    : data{_mm512_set1_pd(x)}
```
Because I changed the constructors for Tensor<rank, dim, Number>, they now call the one-parameter constructor VectorizedArray(double) instead of the assignment operator.

Based on what I remember, when I added the one-parameter constructor, the compiler demanded me to add a default constructor as well.

The generic VectorizedArray<Number> defined in the beginning, on the other hand, didn't need any constructor because it does not need to call any functions like _mm512_set1_pd.

masterleinad commented:
Thanks! Would you mind to rebase as well?

tamiko commented on cmake/checks/check_01_cxx_features.cmake:
```diff
@@ -753,6 +754,19 @@ CHECK_CXX_SOURCE_COMPILES(
+# it will prevent breaking builds with certain compilers.
+#
+IF (DEAL_II_HAVE_CXX14_CONSTEXPR_CAN_CALL_NONCONSTEXPR)
+  MESSAGE(STATUS "Enabling CXX14 constexpr features")
```
Let's remove the status message here. It is a minor detail and will be present in the `config.h` file in case someone wants to look it up.

masterleinad commented:
I rebased and addressed @tamiko's comment.

rezarastak commented:
Sorry I have been busy in the last couple of weeks. Thank you so much for making the necessary changes.

masterleinad commented:
Would you mind turning the one example you described in your first commit into a test case?

rezarastak commented:
Yes I will add that test today. I will add it to the same test file `constexpr_tensor.cc`.

Should I move the test file from `tests/base/` to `tests/tensors/`?

rezarastak commented on include/deal.II/base/tensor.h:
```diff
@@ -1094,9 +1110,14 @@ namespace internal
+    struct Uninitialized
+    {
+      static ArrayElementType value;
+    };
```
Originally, the following function definition existed.
```c++
     template <typename ArrayElementType>
     ArrayElementType &
     subscript(ArrayElementType *,
               const unsigned int,
               std::integral_constant<int, 0>)
     {
       Assert(
         false,
         ExcMessage(
           "Cannot access elements of an object of type Tensor<rank,0,Number>."));
       static ArrayElementType t;
       return t;
     }
```
The static variable defined within the function did now allow constexpr evaluation of the function (see [here](https://godbolt.org/z/ftsoMC)). So I made a workaround by creating `struct Uninitialized` outside the function within the `internal::TensorSubscriptor` namespace.

jppelteret replied:
If this is the case, then your workaround seems reasonable to me. It would be helpful just to document in the code why the struct is necessary. Perhaps a slight permutation of exactly what you've written here, placed as a normal comment just above the struct declaration, would suffice :-)

rezarastak replied:
Added the explanation comment in my latest commit

rezarastak commented:
test added

kronbichler commented on include/deal.II/base/vectorization.h:
```diff
@@ -379,7 +381,7 @@ class VectorizedArray
    * declared public.
    */
-  Number data;
+  Number data = 0.;
```
Can you explain why you need this change (and for the various specializations below)? This is a fundamental change in behavior:
- It is a performance regression because I don't think the compiler will be allowed to skip initialization of `VectorizedArray` to zero when e.g. creating a fixed-size array on the stack.
- This makes this class behave differently than the basic types `double` because a plain variable is initialized, which is undesired.

Shouldn't everything you need be covered by the non-default constructor that you added?

rezarastak replied:
Initially I added these member initialization so that I can remove the default constructor, but we added back all of the constructors and these direct member initialization are not needed.

I was not aware of the performance implications of this. So now I removed member initialization in df218cb. The tests are passing on my machine.

masterleinad commented:
I am seeing these warnings
```
/home/darndt/dealii-new/include/deal.II/base/vectorization.h:2301:7: note: ‘class dealii::VectorizedArray<double>’ has no user-provided default constructor
 2301 | class VectorizedArray<double>
      |       ^~~~~~~~~~~~~~~~~~~~~~~
/home/darndt/dealii-new/include/deal.II/base/vectorization.h:2312:3: note: constructor is not user-provided because it is explicitly defaulted in the class body
 2312 |   VectorizedArray() = default;
      |   ^~~~~~~~~~~~~~~
/home/darndt/dealii-new/include/deal.II/base/vectorization.h:2493:11: note: and the implicitly-defined constructor does not initialize ‘__m128d dealii::VectorizedArray<double>::data’
 2493 |   __m128d data;
```

masterleinad commented:
I pushed a fix for the warning.

kronbichler commented on include/deal.II/base/vectorization.h:
```diff
@@ -180,8 +180,10 @@ class VectorizedArray
+  /**
+   * Default constructor
+   */
+  VectorizedArray() = default;
```
Don't we still need the constructor
```
DEAL_II_ALWAYS_INLINE explicit VectorizedArray(const Number x)
  : data(x)
{}
```

rezarastak replied:
On my machine, if I compile the code, it would never instantiate the general struct `VectorizedArray<Number>`. It always instantiates the specialized versions. So I can't completely test whether we need these constructors or not.

Having said that, yes I think we need to add it.

kronbichler replied:
You can try with `Tensor<1,dim,VectorizedArray<unsigned int>>` or `Tensor<1,dim,VectorizedArray<long double>>`, these two do not go into the specialized versions.

masterleinad commented:
Actually, all `vectorization` and `tensor` tests are passing for me if `vectorization.h` is not modified at all in this pull request.

rezarastak commented:
`vectorization.h` has to be modified in this PR because I added the following constructor that will call the constructor of `VectorizedArray<double>` for `Number = VectorizedArray<double>` and `OtherNumber = double`
```c++
constexpr Tensor<0, dim, Number>::Tensor(const OtherNumber &initializer)
  : value(internal::NumberType<Number>::value(initializer))
{}
```
So VectorizedArray<double> must have a constructor that converts double to the appropriate vectorized variable. such as 
```c++
DEAL_II_ALWAYS_INLINE explicit VectorizedArray(const double x)
    : data{_mm256_set1_pd(x)}
{}
```
which didn't exit before this PR.

masterleinad commented:
> `vectorization.h` has to be modified in this PR because I added the following constructor that will call the constructor of `VectorizedArray<double>` for `Number = VectorizedArray<double>` and `OtherNumber = double`
> 
> ```c++
> constexpr Tensor<0, dim, Number>::Tensor(const OtherNumber &initializer)
>   : value(internal::NumberType<Number>::value(initializer))
> {}
> ```

But the return value of `internal::NumberType<VectorizedArray<Number>>::value(const OtherNumber &)` is `VectorizedArray<Number>` and not `Number`. Have a look at https://godbolt.org/z/SRcEtI.
With the additional constructor `internal::NumberType<VectorizedArray<Number>>::value` can be simplified though.

rezarastak commented:
Yes you are right. We don't need the new constructors for vectorization because `internal::NumberType<VectorizedArray<Number>>::value` doesn't call the constructor, but it uses the assignment operator to initialize `VectorizedArray<Number> tmp`. Therefor, we can revert `vectorization.h`. Then we don't need to worry about any possible performance regressions. I tested it on my workstation too.

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -20,6 +20,12 @@ DEAL_II_NAMESPACE_OPEN
 const unsigned int VectorizedArray<double>::n_array_elements;
 const unsigned int VectorizedArray<float>::n_array_elements;
+static_assert(
+  std::is_standard_layout<VectorizedArray<double>>::value,
```
```suggestion
  std::is_pod<VectorizedArray<double>>::value,
```

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -20,6 +20,12 @@ DEAL_II_NAMESPACE_OPEN
+  std::is_standard_layout<VectorizedArray<double>>::value,
+  "VectorizedArray<double> must have standard (C compatible) layout");
+static_assert(
+  std::is_standard_layout<VectorizedArray<float>>::value,
```
```suggestion
  std::is_pod<VectorizedArray<float>>::value,
```

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -20,6 +20,12 @@ DEAL_II_NAMESPACE_OPEN
 const unsigned int VectorizedArray<float>::n_array_elements;
+static_assert(
+  std::is_standard_layout<VectorizedArray<double>>::value,
+  "VectorizedArray<double> must have standard (C compatible) layout");
```
```suggestion
  "VectorizedArray<double> must be a POD type.");
```

masterleinad commented on source/base/vectorization.cc:
```diff
@@ -20,6 +20,12 @@ DEAL_II_NAMESPACE_OPEN
+  "VectorizedArray<double> must have standard (C compatible) layout");
+static_assert(
+  std::is_standard_layout<VectorizedArray<float>>::value,
+  "VectorizedArray<float> must have standard (C compatible) layout");
```
```suggestion
  "VectorizedArray<float> must be a POD type.");
```

rezarastak commented:
[is_pod is deprecated in c++20](https://en.cppreference.com/w/cpp/types/is_pod). So I think we need to replace it with `is_standard_layout<T> && is_trivial<T>`.

[more info](https://stackoverflow.com/questions/6496545/trivial-vs-standard-layout-vs-pod)

masterleinad commented:
> So I think we need to replace it with `is_standard_layout<T> && is_trivial<T>`.

Sure. That's fine with me.

rezarastak commented:
/rebuild

masterleinad commented:
/rebuild

masterleinad commented:
Thanks, @rezarastak!

rezarastak commented:
Great!
